[Form] Make sure errors are a part of the label on bootstrap 4 - this is a requirement for WCAG2

Nyholm opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | 
| License       | MIT
| Doc PR        | symfony/symfony-docs#... <!--highly recommended for new features-->

I recently let Europe's leading accessibility experts (Funkanu.se) review a site of mine, they gave me (among other) the feedback that errors should be a part of the label. 

They said that it makes no sense for blind users to read label, read input and then read errors. 

I know the implementation might look strange. But I wish something like this would be merged. That would be great for accessibility for all apps using Symfony. 

We *could* also make sure it prints something like:

```
<label for=”name”>Name: <span class=”hidden”>Error message</span></label>
<input id=”name” type=”text”>
<span aria-hidden=”true”>Error message</span>
```

Nyholm committed:
[Form] Make sure errors are a part of the label on bootstrap 4 - this is a requirement for WCAG2

fabpot commented:
Is it really something that we could only implement for Bootstrap 4? Wouldn't it make sense to make the change to the other themes? Also, for Bootstrap 4, it should be done on 3.4, not master.

ostrolucky commented:
Wouldn't changing this in older themes be BC break?
Bootstrap 4 theme has been added in Symfony 3.4 so it's ok to do this there.

fabpot commented:
@ostrolucky I know that, but it's kind of weird to fix a real issue only for one theme.

nicolas-grekas commented:
@Nyholm rebase on 3.4 needed (I already retargeted on github)

nicolas-grekas commented:
@Nyholm I just rebased on 3.4. Can you please review other comments?

Nyholm commented:
Thank you for reviews and comments. I'll take care of this tonight. 

//Tobias Nyholm

Skickat på språng

> On 12 Oct 2017, at 09:16, Nicolas Grekas <notifications@github.com> wrote:
> 
> @Nyholm I just rebased on 3.4. Can you please review other comments?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

Nyholm commented:
I've updated all the themes. 
Please note that the design (depending on CSS) might break with these fixes. 

If we are happy with the implementation I'll try to write a paragraph or two in the docs about this and how to remove the error from the label.

xabbuh commented:
tests seem to need an update too

Nyholm commented:
I updated some tests. It passes locally. Travis is not all happy, did I do it correctly?

nicolas-grekas commented:
@Nyholm could rebase + check tests please?

Nyholm commented:
I've rebased and fixed one test.. I cannot figure out how to fix `AbstractTableLayoutTest::testNestedFormError`. Where is the documentation of DomCrawler?

Nyholm commented on src/Symfony/Component/Form/Tests/AbstractTableLayoutTest.php:
```diff
@@ -327,7 +331,7 @@ public function testNestedFormError()
     [
-        ./tr/td/ul[./li[.="[trans]Error![/trans]"]]
+        ./tr/td/label/ul[./li[.="[trans]Error![/trans]"]]
         /following-sibling::table[@id="name_child"]
```
This line is wrong. I cannot figure out how to make it right. It should match: 

```html
<table id="my&amp;id" class="my&amp;class">
<tr>
        <td><label class="required">[trans]Child[/trans]<ul><li>[trans]Error![/trans]</li></ul></label></td>
        <td><table id="name_child">
<tr>
        <td><label class="required">[trans]Grand child[/trans]</label></td>
        <td><table id="name_child_grandChild">
    </table></td>
    </tr>
    </table></td>
    </tr>
<tr style="display: none">
        <td colspan="2"><input type="hidden" id="name__token" name="name[_token]"></td>
    </tr>
    </table>

```

xabbuh commented:
@Nyholm Tests for the TwigBridge are green now. But it looks like you need to update some PHP templates too (see the failing FrameworkBundle tests).

Nyholm commented:
Tests are green except for low/high. Also, Fabbot is a false negative

Tobion commented on src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_4_layout.html.twig:
```diff
@@ -125,8 +125,8 @@
     {%- else -%}
-        {%- set label_attr = label_attr|merge({class: (label_attr.class|default('') ~ ' form-control-label')|trim}) -%}
-    {%- endif -%}
+    {%- set label_attr = label_attr|merge({class: (label_attr.class|default('') ~ ' form-control-label')|trim}) -%}
```
the indention is broken

Tobion commented:
Don't merge such a bc break in a minor version

nicolas-grekas commented:
@Nyholm how could we move forward here?

ostrolucky commented:
I think this is still ok to go into bootstrap 4 theme, since that one needs to be changed in BC breaking way anyway to support newer bootstrap 4 release

nicolas-grekas commented:
@Nyholm because we cannot break BC, except on bootstrap 4 because it was not even stable before, what about focusing this PR only on bootstrap 4? That'd be a nice step. Would you have some time to make it before next release? (ie this week would be great)

Nyholm commented:
Hey. Sorry for the delay. I will make sure to only change the bootstrap theme this week.

Nyholm commented:
There we go. The PR does only edit bootstrap 4 and added some tests to make sure it works. 

I've also rebased on branch 3.4. But I guess it should be changed to master, right?

nicolas-grekas commented:
This should realy go to 3.4 as bug fix. Later would mean BC breaks. Doing now is also a BC break, but matches the fact that BS4 just when out of beta.

Nyholm commented:
Great great great. Then Im done with this PR. Im ready for a review.

nicolas-grekas commented on src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_4_layout.html.twig:
```diff
@@ -39,8 +39,42 @@
+            {%- if with_invert %}{{ form_widget(form.invert) }}{% endif -%}
+        </div>
+    {%- endif -%}
+{%- endblock dateinterval_widget -%}
```
the last dash could be removed, just to remove an unneeded change in the patch

nicolas-grekas commented on src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_4_layout.html.twig:
```diff
@@ -125,13 +159,28 @@
+            {%- set element = 'legend' -%}
+            {%- set label_attr = label_attr|merge({class: (label_attr.class|default('') ~ ' col-form-legend')|trim}) -%}
+        {%- else -%}
+            {%- set label_attr = label_attr|merge({'for': id, class: (label_attr.class|default('') ~ ' form-control-label')|trim}) -%}
```
no need for quotes around key (same below)

nicolas-grekas reviewed (approved):
(with minor nitpicking comments)

Nyholm commented:
Thank you

nicolas-grekas commented:
Thank you @Nyholm.

Nyholm commented:
Thank you for merging.
